doc: onNavigate (#77647)
vercel/next.js: docs/01-app/04-api-reference/02-components/link.mdx

<PagesOnly>

| Prop                                | Example                  | Type              | Required |
| ----------------------------------- | ------------------------ | ----------------- | -------- |
| [`href`](#href-required)            | `href="/dashboard"`      | String or Object  | Yes      |
| [`replace`](#replace)               | `replace={false}`        | Boolean           | -        |
| [`scroll`](#scroll)                 | `scroll={false}`         | Boolean           | -        |
| [`prefetch`](#prefetch)             | `prefetch={false}`       | Boolean           | -        |
| [`legacyBehavior`](#legacybehavior) | `legacyBehavior={true}`  | Boolean           | -        |
| [`passHref`](#passhref)             | `passHref={true}`        | Boolean           | -        |
| [`shallow`](#shallow)               | `shallow={false}`        | Boolean           | -        |
| [`locale`](#locale)                 | `locale="fr"`            | String or Boolean | -        |

> | [`onNavigate`](#onnavigate)         | `onNavigate={(e) => {}}` | Function          | -        |

Reviewer: To confirm, this feature works for Pages and App Router?

Author: Yep

Author: Yep

### `onNavigate`

An event handler called during client-side navigation. The handler receives an event object that includes a `preventDefault()` method, allowing you to cancel the navigation if needed.

```tsx filename="app/page.tsx" switcher
import Link from 'next/link'

export default function Page() {
  return (
    <Link
      href="/dashboard"
      onNavigate={(e) => {
        // Only executes during SPA navigation
        console.log('Navigating...')

        // Optionally prevent navigation
        // e.preventDefault()
      }}
    >
      Dashboard
    </Link>
  )
}
```

```jsx filename="app/page.js" switcher
import Link from 'next/link'

export default function Page() {
  return (
    <Link
      href="/dashboard"
      onNavigate={(e) => {
        // Only executes during SPA navigation
        console.log('Navigating...')

        // Optionally prevent navigation
        // e.preventDefault()
      }}
    >
      Dashboard
    </Link>
  )
}
```

> **Good to know**: While `onClick` and `onNavigate` may seem similar, they serve different purposes. `onClick` executes for all click events, while `onNavigate` only runs during client-side navigation. Some key differences:
>

> > - When using modifier keys (Ctrl/Cmd+Click), `onClick` executes but `onNavigate` doesn't since Next.js prevents default navigation for new tabs.

Reviewer 2: ```suggestion
> - When using modifier keys (`Ctrl`/`Cmd` + Click), `onClick` executes but `onNavigate` doesn't since Next.js prevents default navigation for new tabs.
```

> > - External URLs won't trigger `onNavigate` since it's only for SPA-like navigations.

Reviewer 2: ```suggestion
> - External URLs won't trigger `onNavigate` since it's only for client-side and same-origin navigations.
```

> > - Links with the `download` attribute will call `onClick` but not `onNavigate` since downloads use browser navigation.

Reviewer 2: ```suggestion
> - Links with the `download` attribute will work with `onClick` but not `onNavigate` since the browser will treat the linked URL as a download. 
```

</PagesOnly>

### Blocking navigation

> You can use the `onNavigate` prop to block navigation when certain conditions are met, such as when a form has unsaved changes. When you need to block navigation across multiple components in your app (like preventing navigation from any link while a form is being edited), React Context provides a clean way to share this blocking state.
> 
> Here's how to implement this pattern:
> 
> First, create a context to track the navigation blocking state:

Reviewer 2: Remove the double :

```suggestion
You can use the `onNavigate` prop to block navigation when certain conditions are met, such as when a form has unsaved changes. When you need to block navigation across multiple components in your app (like preventing navigation from any link while a form is being edited), React Context provides a clean way to share this blocking state. First, create a context to track the navigation blocking state:
```

  return (
    <nav>
      <Link href="/">Home</Link>
      <Link href="/about">About</Link>
    </nav>
  )
}
```

```jsx filename="app/components/nav.js" switcher
'use client'

import { CustomLink as Link } from './custom-link'

export default function Nav() {
  return (
    <nav>
      <Link href="/">Home</Link>
      <Link href="/about">About</Link>
    </nav>
  )
}
```

Finally, wrap your app with the `NavigationBlockerProvider` in the root layout and use the components in your page:

```tsx filename="app/layout.tsx" switcher
import { NavigationBlockerProvider } from './contexts/navigation-blocker'

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html lang="en">
      <body>
        <NavigationBlockerProvider>{children}</NavigationBlockerProvider>
      </body>
    </html>
  )
}
```

```jsx filename="app/layout.js" switcher
import { NavigationBlockerProvider } from './contexts/navigation-blocker'

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>
        <NavigationBlockerProvider>{children}</NavigationBlockerProvider>
      </body>
    </html>
  )
}
```

> Then use the Nav and Form components in your page:

Reviewer 2: ```suggestion
Then, use the `Nav` and `Form` components in your page:
```

```tsx filename="app/page.tsx" switcher
import Nav from './components/nav'
import Form from './components/form'

export default function Page() {
  return (
    <div>
      <Nav />
      <main>
        <h1>Welcome to the Dashboard</h1>
        <Form />
      </main>
    </div>
  )
}
```

```jsx filename="app/page.js" switcher
import Nav from './components/nav'
import Form from './components/form'

export default function Page() {
  return (
    <div>
      <Nav />
      <main>
        <h1>Welcome to the Dashboard</h1>
        <Form />
      </main>
    </div>
  )
}
```

> Now when a user tries to navigate away using any `CustomLink` while the form has unsaved changes, they'll be prompted to confirm before leaving.

Reviewer 2: ```suggestion
When a user tries to navigate away using `CustomLink` while the form has unsaved changes, they'll be prompted to confirm before leaving.
```

| Version   | Changes                                                                                                                                                                                         |
| --------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `v15.3.0` | Add `onNavigate`                                                                                                                                                                                |

Reviewer: ```suggestion
| `v15.3.0` | Add `onNavigate` API                                                                                                                                                                               |
```
